### PR TITLE
[RDY] Enable -Wconversion in ops.c

### PR DESCRIPTION
**Pull request by cjoach**
Work for #567
~~Can someone help me with the build error?
The assert that fails on line 1431 checks the explicit conversion that was implicit on line 1432. 
Also, I don't understand what to do with the clang address sanitizer problem.
I'll look shortly into the 32bit test that fails but if anyone has a clue on it I'd be glad.~~

Edit: Waiting for comments

### CODE REVIEW

File: src/nvim/ops.c [hunk 1]
```diff
@@ -2345,9 +2349,11 @@ static void op_yank_reg(oparg_T *oap, bool message, yankreg_T *reg, bool append)
   }
 
   reg->y_size = yanklines;
-  reg->y_type = yanktype;     /* set the yank register type */
+  assert(yanktype <= UCHAR_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: as yanktype is not really a numerical quantity but essentially an enum (one of MLINE, MCHAR, MBLOCK, MAUTO), this assert doesn't really make sense. I think you can just drop this one, but if an assert is added for motiontype/yanktype anywhere, it is better to tightly check that is is one of the 4 (or sometimes less) legal values.

File: src/nvim/ops.c [hunk 2]
```diff
@@ -2440,7 +2446,8 @@ static void op_yank_reg(oparg_T *oap, bool message, yankreg_T *reg, bool append)
   }
 
   if (curr != reg) {      /* append the new block to the old block */
-    new_ptr = xmalloc(sizeof(char_u *) * (curr->y_size + reg->y_size));
+    new_ptr = xmalloc(sizeof(char_u *)
+                      * (unsigned long)(curr->y_size + reg->y_size));
```
Comment (bfredl) [inline src/nvim/ops.c]: why not `(size_t)` ?

File: src/nvim/normal.c [hunk 3]
```diff
@@ -2168,13 +2168,12 @@ static void op_function(oparg_T *oap)
  *
  * Return true if start_arrow() should be called for edit mode.
  */
-bool
```
Comment (bfredl) [inline src/nvim/normal.c]: To make PRs easier to review, please try to avoid to make completely unrelated changes in a single PR (this seems to have nothing to do with `ops.c`)

Also, when you reformat a signature (which is ok if you do a related change to the function), please reformat to doxygen style, like
```
/// Yanks the text between "oap->start" and "oap->end" into a yank register.
/// If we are to append (uppercase register), we first yank into a new yank
/// register and then concatenate the old and the new one.
///
/// @param oap operator arguments
/// @param message show message when more than `&report` lines are yanked.
/// @returns whether the operation register was writable.
bool op_yank(oparg_T *oap, bool message)
```
i e param comments are before, and then the signature itself typically fits on one line.

File: src/nvim/ops.c [hunk 4]
```diff
@@ -455,13 +454,12 @@ static void block_insert(oparg_T *oap, char_u *s, int b_insert, struct block_def
   int count = 0;                /* extra spaces to replace a cut TAB */
   int spaces = 0;               /* non-zero if cutting a TAB */
   colnr_T offset;               /* pointer along new line */
-  unsigned s_len;               /* STRLEN(s) */
+  size_t s_len = STRLEN(s);     // STRLEN(s)
```
Comment (bfredl) [inline src/nvim/ops.c]: with this change the comment gets  redundant

File: src/nvim/ops.c [hunk 5]
```diff
@@ -516,8 +515,9 @@ static void block_insert(oparg_T *oap, char_u *s, int b_insert, struct block_def
     memset(newp + offset, ' ', (size_t)spaces);
 
     /* copy the new text */
-    memmove(newp + offset + spaces, s, (size_t)s_len);
-    offset += s_len;
+    memmove(newp + offset + spaces, s, s_len);
+    assert(s_len <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: assert can be removed, as inserted text `s` (inserted by the user on the first line of the block) couldn't be longer than MAXCOL to start with.

File: src/nvim/ops.c [hunk 6]
```diff
@@ -787,7 +787,8 @@ yankreg_T *copy_register(int name)
   if (copy->y_size == 0) {
     copy->y_array = NULL;
   } else {
-    copy->y_array = xcalloc(copy->y_size, sizeof(char_u *));
+    assert(copy->y_size >= 0);
```
Comment (bfredl) [inline src/nvim/ops.c]: Unless we check every time we read `yankreg_T->y_size` I don't think this really helps with robustness. If we feel the need, It is better to assert when writing y_size (when it is the result of a nontrivial calculation), which you already do in some place.

File: src/nvim/ops.c [hunk 7]
```diff
@@ -1028,7 +1028,8 @@ static void put_reedit_in_typebuf(int silent)
       buf[1] = 'R';
       buf[2] = NUL;
     } else {
-      buf[0] = restart_edit == 'I' ? 'i' : restart_edit;
+      assert(restart_edit <= UCHAR_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: not needed as restart_edit is only ever set to 0 or character constant. It is probably better (in a later PR) to change its type to char or char_u.

File: src/nvim/ops.c [hunk 8]
```diff
@@ -1425,7 +1422,7 @@ int op_delete(oparg_T *oap)
        */
       n = bd.textlen - bd.startspaces - bd.endspaces;
       oldp = ml_get(lnum);
-      newp = (char_u *) xmalloc((size_t)(STRLEN(oldp) + 1 - n));
+      newp = (char_u *)xmalloc((size_t)((int)STRLEN(oldp) + 1 - n));
```
Comment (bfredl) [inline src/nvim/ops.c]: is it not simpler to convert `n` to `(size_t)n`

File: src/nvim/ops.c [hunk 9]
```diff
@@ -2197,14 +2197,15 @@ int op_change(oparg_T *oap)
       long new_indent = (long)(skipwhite(firstline) - firstline);
 
       pre_textlen += new_indent - pre_indent;
-      bd.textcol += new_indent - pre_indent;
+      assert(new_indent - pre_indent <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: what does this assert against?

File: src/nvim/ops.c [hunk 10]
```diff
@@ -2217,10 +2218,13 @@ int op_change(oparg_T *oap)
           if (bd.is_short) {
             vpos.lnum = linenr;
             (void)getvpos(&vpos, oap->start_vcol);
-          } else
+          } else {
             vpos.coladd = 0;
+          }
           oldp = ml_get(linenr);
-          newp = (char_u *) xmalloc((size_t)(STRLEN(oldp) + vpos.coladd + ins_len + 1));
+          assert(ins_len >= 0);
```
Comment (bfredl) [inline src/nvim/ops.c]: superfluous (we are inside `if (ins_len > 0)` )

File: src/nvim/ops.c [hunk 11]
```diff
@@ -4376,7 +4384,8 @@ int do_addsub(int command, linenr_T Prenum1)
       --length;
     }
     if (hex == 'x' || hex == 'X') {
-      *ptr++ = hex;
+      assert(hex >= 0 && hex <= UCHAR_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: not needed (surrounding `if()` )

File: src/nvim/ops.c [hunk 12]
```diff
@@ -4643,15 +4654,11 @@ void write_reg_contents_lst(int name, char_u **strings, int maxlen,
 /// @param block_len width of visual block
 void write_reg_contents_ex(int name,
                            const char_u *str,
-                           ssize_t len,
+                           size_t len,
                            bool must_append,
                            int yank_type,
                            long block_len)
 {
-  if (len < 0) {
```
Comment (bfredl) [inline src/nvim/ops.c]: it seems you have handled all cases this affects, but I'm not sure we really want this. `-1` for an unlimited length seems to be a common idiom in the codebase, and I don't know if we want to get rid of it. Anyway it seems unrelated to enabling `-Wconversion` ?

File: src/nvim/ops.c [hunk 13]
```diff
@@ -4704,7 +4711,7 @@ void write_reg_contents_ex(int name,
 /// @param blocklen width of visual block, or -1 for "I don't know."
 /// @param str_list True if str is `char_u **`.
 static void str_to_reg(yankreg_T *y_ptr, int yank_type, const char_u *str,
-                       size_t len, colnr_T blocklen, bool str_list)
+                       size_t len, long blocklen, bool str_list)
```
Comment (bfredl) [inline src/nvim/ops.c]: what does this change help with?

File: src/nvim/ops.c [hunk 14]
```diff
@@ -5133,7 +5145,8 @@ static bool get_clipboard(int name, yankreg_T **target, bool quiet)
   free_register(reg);
 
   list_T *args = list_alloc();
-  char_u regname = name;
+  assert(name <= UCHAR_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: not needed (regnames are asserted ASCII earlier in `valid_yank_reg` )

File: src/nvim/ops.c [hunk 15]
```diff
@@ -5272,7 +5287,8 @@ static void set_clipboard(int name, yankreg_T *reg)
   }
   list_append_string(args, &regtype, 1);
 
-  char_u regname = name;
+  assert(name <= UCHAR_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: as above

File: src/nvim/buffer_defs.h [hunk 16]
```diff
@@ -649,7 +649,7 @@ struct file_buffer {
   bool b_p_swf;                 /* 'swapfile' */
   long b_p_smc;                 /* 'synmaxcol' */
   char_u      *b_p_syn;         /* 'syntax' */
-  long b_p_ts;                  /* 'tabstop' */
+  int b_p_ts;                   // 'tabstop'
```
Comment (bfredl) [inline src/nvim/buffer_defs.h]: do we already have any numeric option that is not `long` ? If not then I don't think we can be sure that it is safe, for instance in `option.c` all numeric options are read/written as `*(long*)`.

File: src/nvim/ops.c (hunk 12 above)
Comment (cacplate) [inline src/nvim/ops.c]: it is indeed a common idiom. I don't know where I got the idea but I thought that we indeed wanted to get rid of it. Maybe @justinmk can light us on that. If I must revert, do you think it's best to revert all cases or just to handle again the `-1`?

File: src/nvim/buffer_defs.h (hunk 16 above)
Comment (cacplate) [inline src/nvim/buffer_defs.h]: it creates a great deal of warnings if I keep it to long. See #3803 for my change of numerical options to int. There is no problem in this PR for now but indeed this is not ideal and the changes in window.c will hopefully change that if it's deemed good enough.

Comment (bfredl) [inline src/nvim/buffer_defs.h]: ok, should be fine then in the meanwhile  if we can accept all num options eventually becoming `int` values. (we actually have the same inconsistency for boolean options already, I will look into that case as soon as I have a bit more time).

File: src/nvim/message.c [hunk 17]
```diff
@@ -2449,8 +2449,13 @@ static void redir_write(char_u *str, int maxlen)
       }
     }
 
-    if (redir_reg)
-      write_reg_contents(redir_reg, s, maxlen, TRUE);
+    if (redir_reg) {
+      if (maxlen == -1) {
```
Comment (bfredl) [inline src/nvim/message.c]: I think 
`size_t len = maxlen == -1 ? STRLEN(s) : maxlen;`
is clearer (DRY)

File: src/nvim/message.c [hunk 18]
```diff
@@ -468,7 +468,7 @@ int emsg_not_now(void)
  *
  * return TRUE if wait_return not called
  */
-int emsg(char_u *s)
```
Comment (bfredl) [inline src/nvim/message.c]: unrelated (this belongs in a PR for message.c)

File: src/nvim/normal.h [hunk 19]
```diff
@@ -36,8 +36,8 @@ typedef struct oparg_S {
   bool block_mode;              /* current operator is Visual block mode */
   colnr_T start_vcol;           /* start col for block mode operator */
   colnr_T end_vcol;             /* end col for block mode operator */
-  long prev_opcount;            // ca.opcount saved for K_EVENT
```
Comment (bfredl) [inline src/nvim/normal.h]: are these (and `cmdarg_T` below) needed to fix `-Wconversion` in `ops.c` ?

File: src/nvim/ops.c [hunk 20]
```diff
@@ -1169,8 +1162,7 @@ static void stuffescaped(char_u *arg, int literally)
  * If "regname" is a special register, return TRUE and store a pointer to its
  * value in "argp".
  */
-int 
-get_spec_reg (
+int get_spec_reg(
```
Comment (bfredl) [inline src/nvim/ops.c]: In future PRs please avoid style changes in functions that otherwise are unrelated to the subject of the PR, or at least in a separate commit so that semantic changes (integer types and casts in this case) can be reviewed separately from the stylistic ones. (small changes close to lines that are changed anyway are of course ok, and sometimes required by the linter)

Comment (justinmk) [inline src/nvim/ops.c]: AFAIK our linter is aggressively nudging people to fix nearby style issues.

Comment (bfredl) [inline src/nvim/ops.c]: sorry, I thought linter would only affected the closest 2-3 lines or so to a change and not a completely unchanged part of the file, but maybe that changed in the meanwhile (I was too busy the last 1-2 months to really follow all nvim discussions)

Comment (justinmk) [inline src/nvim/ops.c]: You're correct, I didn't check that.

File: src/nvim/message.c [hunk 21]
```diff
@@ -2438,7 +2438,7 @@ static void redir_write(char_u *str, int maxlen)
     if (*s != '\n' && *s != '\r') {
       while (cur_col < msg_col) {
         if (redir_reg)
-          write_reg_contents(redir_reg, (char_u *)" ", -1, TRUE);
+          write_reg_contents(redir_reg, (char_u *)" ", STRLEN(" "), true);
```
Comment (ZyX-I) [inline src/nvim/message.c]: For static strings there is `sizeof(" ") - 1`. Though I have no idea why you can’t just write `1` here, I would prefer this. (Also no need to use `STRLEN` in case when argument is `char*`: do not know others’ opinion, but I tend to avoid uneeded macros and use `strlen`.)

File: src/nvim/buffer_defs.h [hunk 22]
```diff
@@ -649,7 +649,7 @@ struct file_buffer {
   bool b_p_swf;                 /* 'swapfile' */
   long b_p_smc;                 /* 'synmaxcol' */
   char_u      *b_p_syn;         /* 'syntax' */
```
Comment (ZyX-I) [inline src/nvim/buffer_defs.h]: You will clearly have to convert *all* comments here from `/*` to `//`: adjacent lines, linter complains. Better in a separate commit. Also making them look like `///< 'syntax' option value` is probably more appropriate.

File: src/nvim/ex_docmd.c [hunk 23]
```diff
@@ -7238,7 +7238,7 @@ static void ex_redir(exarg_T *eap)
           /* Make register empty when not using @A-@Z and the
            * command is valid. */
           if (*arg == NUL && !isupper(redir_reg))
-            write_reg_contents(redir_reg, (char_u *)"", -1, FALSE);
+            write_reg_contents(redir_reg, (char_u *)"", STRLEN(""), false);
```
Comment (ZyX-I) [inline src/nvim/ex_docmd.c]: 0, `sizeof`, `strlen`.

File: src/nvim/normal.c [hunk 24]
```diff
@@ -2132,50 +2132,47 @@ static void op_function(oparg_T *oap)
...
+///
+/// (1) only if mouse pointer moved since press
+/// (2) only if click is in same buffer
+///
+/// @param oap operator argument, can be NULL
+/// @param c K_LEFTMOUSE, etc
+/// @param dir Direction to 'put' if necessary
+/// @param count
+/// @param fixindent PUT_FIXINDENT if fixing indent necessary
+//
```
Comment (ZyX-I) [inline src/nvim/normal.c]: `///`

File: src/nvim/ops.c (hunk 20 above)
Comment (cacplate) [inline src/nvim/ops.c]: this is indeed my doing and not the linter's idea. I'll refrain from doing this sort of change in the future.

File: src/nvim/normal.h (hunk 19 above)
Comment (cacplate) [inline src/nvim/normal.h]: yes. either this or casting and it doesn't seem to create problems.

File: src/nvim/message.c (hunk 21 above)
Comment (cacplate) [inline src/nvim/message.c]: noted for the strlen, I think I also prefer no macro if not necessary. I will fix this also.

File: src/nvim/buffer_defs.h (hunk 22 above)
Comment (cacplate) [inline src/nvim/buffer_defs.h]: I though I'd do linting after review. Do you think I'm better doing multiple commits to split linting?

Comment (ZyX-I) [inline src/nvim/buffer_defs.h]: Nope. Review needs to review *all* changes. It is also OK if the only commit that passes linting is the last one.

File: src/nvim/buffer_defs.h (hunk 16 above)
Comment (cacplate) [inline src/nvim/buffer_defs.h]: I will revert this change and cast to dissociate this PR from the "numerical options" problem so this PR can be merged independently.

File: src/nvim/ex_cmds2.c [hunk 25]
```diff
@@ -1867,9 +1867,7 @@ void ex_argdelete(exarg_T *eap)
   maketitle();
 }
 
-/*
- * ":argdo", ":windo", ":bufdo", ":tabdo", ":cdo", ":ldo", ":cfdo" and ":lfdo"
- */
+/// ":argdo", ":windo", ":bufdo", ":tabdo"
```
Comment (bfredl) [inline src/nvim/ex_cmds2.c]: any reason `:cdo` etc are removed here?

File: src/nvim/globals.h [hunk 26]
```diff
@@ -847,8 +847,8 @@ EXTERN int* (*iconv_errno)(void);
 EXTERN int State INIT(= NORMAL);        /* This is the current state of the
                                          * command interpreter. */
 
-EXTERN int finish_op INIT(= FALSE);     /* TRUE while an operator is pending */
-EXTERN long opcount INIT(= 0);          /* count for pending operator */
+EXTERN int finish_op INIT(= false);     // TRUE while an operator is pending
```
Comment (bfredl) [inline src/nvim/globals.h]: as you had to change this line anyway due to linting, you can just as well update it completely :) (`bool` type and `true` in comment)

File: src/nvim/indent.c [hunk 27]
```diff
@@ -120,7 +120,7 @@ int set_indent(int size, int flags)
       // Count as many characters as we can use.
       while (todo > 0 && ascii_iswhite(*p)) {
         if (*p == TAB) {
-          tab_pad = (int)curbuf->b_p_ts - (ind_done % (int)curbuf->b_p_ts);
```
Comment (bfredl) [inline src/nvim/indent.c]: wasn't all changes in `indent.c` related to the `curbuf->b_p_ts` change? (not saying the changes are wrong, but they are not related to this PR anymore?)

File: src/nvim/misc1.c [hunk 28]
```diff
@@ -185,12 +185,11 @@ open_line (
...
       || do_si
       ) {
-    /*
-     * count white space on current line
-     */
-    newindent = get_indent_str(saved_line, (int)curbuf->b_p_ts, FALSE);
-    if (newindent == 0 && !(flags & OPENLINE_COM_LIST))
-      newindent = second_line_indent;       /* for ^^D command in insert mode */
+    // count white space on current line
+    newindent = get_indent_str(saved_line, curbuf->b_p_ts, false);
```
Comment (bfredl) [inline src/nvim/misc1.c]: the `(int)` should remain here (no warning as `misc1.c` is not `-Wconversion`)

File: src/nvim/misc1.c [hunk 29]
```diff
@@ -625,7 +624,7 @@ open_line (
         if (curbuf->b_p_ai
             || do_si
             )
-          newindent = get_indent_str(leader, (int)curbuf->b_p_ts, FALSE);
+          newindent = get_indent_str(leader, curbuf->b_p_ts, false);
```
Comment (bfredl) [inline src/nvim/misc1.c]: dito

File: src/nvim/option.c [hunk 30]
```diff
@@ -5625,8 +5626,9 @@ void buf_copy_options(buf_T *buf, int flags)
         buf->b_p_isk = save_p_isk;
       else {
         buf->b_p_isk = vim_strsave(p_isk);
-        did_isk = TRUE;
-        buf->b_p_ts = p_ts;
+        did_isk = true;
+        assert(p_ts <= INT_MAX);
+        buf->b_p_ts = (int)p_ts;
```
Comment (bfredl) [inline src/nvim/option.c]: also not needed anymore

File: src/nvim/screen.c [hunk 31]
```diff
@@ -3330,8 +3330,7 @@ win_line (
           // TODO: is passing p for start of the line OK?
           n_extra = win_lbr_chartabsize(wp, line, p, (colnr_T)vcol, NULL) - 1;
           if (c == TAB && n_extra + col > wp->w_width) {
-            n_extra = (int)wp->w_buffer->b_p_ts
-                      - vcol % (int)wp->w_buffer->b_p_ts - 1;
```
Comment (bfredl) [inline src/nvim/screen.c]: revert

File: src/nvim/screen.c [hunk 32]
```diff
@@ -3379,8 +3378,8 @@ win_line (
             vcol_adjusted = vcol - MB_CHARLEN(p_sbr);
           }
           // tab amount depends on current column
-          tab_len = (int)wp->w_buffer->b_p_ts
```
Comment (bfredl) [inline src/nvim/screen.c]: revert

File: src/nvim/ops.c [hunk 33]
```diff
@@ -352,18 +351,18 @@ static void shift_block(oparg_T *oap, int amount)
...
+    colnr_T destination_col;      // column to which text in block will
+                                  // be shifted
+    char_u *verbatim_copy_end;    // end of the part of the line which is
+                                  // copied verbatim
+    colnr_T verbatim_copy_width;  // the (displayed) width of this part
+                                  // of line
+    size_t fill;                  // nr of spaces that replace a TAB
+    size_t new_line_len;          // the length of the line after the
+                                  // block shift
+    int block_space_width;
```
Comment (bfredl) [inline src/nvim/ops.c]: here you are changing to a weaker type (`size_t` -> `int`), shouldn't it be `colnr_T` ?

File: src/nvim/normal.h [hunk 34]
```diff
@@ -10,53 +10,50 @@
...
-
-  long line_count;              /* number of lines from op_start to op_end
-                                   (inclusive) */
-  bool empty;                   /* op_start and op_end the same (only used by
-                                   op_change()) */
-  bool is_VIsual;               /* operator on Visual area */
-  bool block_mode;              /* current operator is Visual block mode */
-  colnr_T start_vcol;           /* start col for block mode operator */
-  colnr_T end_vcol;             /* end col for block mode operator */
-  long prev_opcount;            // ca.opcount saved for K_EVENT
```
Comment (bfredl) [inline src/nvim/normal.h]: Still, I'm not happy with this change in this PR, as `count`s affect much code outside of `ops.c`. (it is also not consistent, there is still count parameters as `long`) Reverting all the changed `count` types back to `long` creates just two warnings in `ops.c`, which should be easy to fix.

File: src/nvim/ex_cmds2.c (hunk 25 above)
Comment (cacplate) [inline src/nvim/ex_cmds2.c]: Oops, refactoring typo I think.

File: src/nvim/indent.c (hunk 27 above)
Comment (cacplate) [inline src/nvim/indent.c]: good point.

File: src/nvim/globals.h [hunk 35]
```diff
@@ -847,8 +847,8 @@ EXTERN int* (*iconv_errno)(void);
 EXTERN int State INIT(= NORMAL);        /* This is the current state of the
                                          * command interpreter. */
 
-EXTERN int finish_op INIT(= FALSE);     /* TRUE while an operator is pending */
-EXTERN long opcount INIT(= 0);          /* count for pending operator */
+EXTERN bool finish_op INIT(= false);    // true while an operator is pending
+EXTERN int opcount INIT(= 0);           // count for pending operator
```
Comment (bfredl) [inline src/nvim/globals.h]: this should be reverted to long as well (`opcount` is not even used in `ops.c`).

File: src/nvim/normal.c [hunk 36]
```diff
@@ -966,7 +966,8 @@ static void normal_finish_command(NormalState *s)
   }
 
   // Save count before an operator for next time
-  opcount = s->ca.opcount;
+  assert(s->ca.opcount <= INT_MAX);
+  opcount = (int)s->ca.opcount;
```
Comment (bfredl) [inline src/nvim/normal.c]: dito

File: src/nvim/normal.c [hunk 37]
```diff
@@ -1414,11 +1415,11 @@ void do_pending_operator(cmdarg_T *cap, int old_col, bool gui_yank)
...
-  /* The visual area is remembered for redo */
-  static int redo_VIsual_mode = NUL;        /* 'v', 'V', or Ctrl-V */
-  static linenr_T redo_VIsual_line_count;   /* number of lines */
-  static colnr_T redo_VIsual_vcol;          /* number of cols or end column */
-  static long redo_VIsual_count;            /* count for Visual operator */
+  // The visual area is remembered for redo
+  static int redo_VIsual_mode = NUL;       ///< 'v', 'V', or Ctrl-V
+  static linenr_T redo_VIsual_line_count;  ///< number of lines
+  static colnr_T redo_VIsual_vcol;         ///< number of cols or end column
+  static int redo_VIsual_count;            ///< count for Visual operator
```
Comment (bfredl) [inline src/nvim/normal.c]: dito

File: src/nvim/globals.h (hunk 35 above)
Comment (cacplate) [inline src/nvim/globals.h]: damn, I lost it. Thank you for your patience.

File: src/nvim/edit.c [hunk 38]
```diff
@@ -8036,12 +8036,13 @@ static int ins_tab(void)
...
-  if (p_sta && ind)             /* insert tab in indent, use 'shiftwidth' */
+  if (p_sta && ind) {  // insert tab in indent, use 'shiftwidth'
     temp = get_sw_value(curbuf);
-  else if (curbuf->b_p_sts != 0)   /* use 'softtabstop' when set */
+  } else if (curbuf->b_p_sts != 0) {  // use 'softtabstop' when set
     temp = get_sts_value();
-  else                          /* otherwise use 'tabstop' */
-    temp = (int)curbuf->b_p_ts;
+  } else {  // otherwise use 'tabstop'
+    temp = curbuf->b_p_ts;
```
Comment (bfredl) [inline src/nvim/edit.c]: sorry missed this one, should also be reverted.

File: src/nvim/ops.c [hunk 39]
```diff
@@ -352,18 +351,18 @@ static void shift_block(oparg_T *oap, int amount)
...
+                                  // be shifted
+    char_u *verbatim_copy_end;    // end of the part of the line which is
+                                  // copied verbatim
+    colnr_T verbatim_copy_width;  // the (displayed) width of this part
+                                  // of line
+    size_t fill;                  // nr of spaces that replace a TAB
+    size_t new_line_len;          // the length of the line after the
+                                  // block shift
+    colnr_T block_space_width;
+    int shift_amount;
```
Comment (bfredl) [inline src/nvim/ops.c]: `colnr_T`

File: src/nvim/ops.c [hunk 40]
```diff
@@ -3535,11 +3534,8 @@ int do_join(long count,
   del_lines(count - 1, FALSE);
   curwin->w_cursor.lnum = t;
 
-  /*
-   * Set the cursor column:
-   * Vi compatible: use the column of the first join
```
Comment (bfredl) [inline src/nvim/ops.c]: have we removed (or decided to) this compatibility option? otherwise this comment should be kept

File: src/nvim/ops.c [hunk 41]
```diff
@@ -4791,7 +4785,7 @@ static void str_to_reg(yankreg_T *y_ptr, int yank_type, const char_u *str,
   set_yreg_additional_data(y_ptr, NULL);
   y_ptr->timestamp = os_time();
   if (type == MBLOCK) {
-    y_ptr->y_width = (blocklen == -1 ? (colnr_T) maxlen - 1 : blocklen);
+    y_ptr->y_width = (colnr_T)(blocklen == -1 ? maxlen - 1 : (size_t)blocklen);
```
Comment (bfredl) [inline src/nvim/ops.c]: I think this was better as before.

File: src/nvim/ops.c (hunk 40 above)
Comment (justinmk) [inline src/nvim/ops.c]: The check for `CPO_JOINCOL` so we didn't remove it at this point.

**Comment by cacplate** [comment]
@bfredl @ZyX-I do you have any other comment? If not, I'll mark this one as `RDY`.

**Comment by bfredl** [comment]
I will take a look tomorrow

File: src/nvim/ops.c [hunk 42]
```diff
@@ -2524,23 +2517,24 @@ static void yank_copy_line(yankreg_T *reg, struct block_def *bd, long y_idx)
...
     dir: BACKWARD for 'P', FORWARD for 'p' */
 void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
 {
-  char_u      *ptr;
-  char_u      *newp, *oldp;
+  char_u *ptr;
+  char_u *newp;
+  char_u *oldp;
   int yanklen;
-  int totlen = 0;                       /* init for gcc */
```
Comment (bfredl) [inline src/nvim/ops.c]: I would like this to be `size_t` but I don't understand under what circumstance `count` can be negative. I _think_ only in the special case `regname == '.'`, at least the rest of the code it looks like `count = -1` has the same effect as `0`.
(even in the `.` case I don't understand, as `".` is _always_ charwise so why would you `o`/`O` it?)

File: src/nvim/ops.c [hunk 43]
```diff
@@ -2927,13 +2926,14 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
      */
     if (y_type == MCHAR && y_size == 1) {
       do {
-        totlen = count * yanklen;
+        assert(count <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: a more relevant assert would be `count >= 0` but as said not 100% sure that would be correct. I suppose we could try it and see if it trigger on travis.

File: src/nvim/ops.c [hunk 44]
```diff
@@ -2874,7 +2872,8 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
 
     /* adjust '] mark */
     curbuf->b_op_end.lnum = curwin->w_cursor.lnum - 1;
-    curbuf->b_op_end.col = bd.textcol + totlen - 1;
+    assert(totlen <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: I don't really like just adding a assert everytime a value is downcast, at least not where there is a more relevant _dynamic_ condition to check, i e `bd.textcol + totlen - 1 < MAXCOL` in this case  (unless this already checked above somewhere...)

File: src/nvim/ops.c (hunk 43 above)
Comment (cacplate) [inline src/nvim/ops.c]: I ran the tests with a `assert(count >= 0)` at the start of the function and nothing happened. This seems to never happen.

File: src/nvim/ops.c (hunk 44 above)
Comment (cacplate) [inline src/nvim/ops.c]: but then the change of `totlen` to `size_t` makes this cast and consequently the assert required (correct me if I'm mistaken).

Comment (bfredl) [inline src/nvim/ops.c]: just change the cast to `(colnr_T)(bd.textcol + totlen - 1)` and it would be fine?

File: src/nvim/ops.c (hunk 43 above)
Comment (bfredl) [inline src/nvim/ops.c]: actually I tried to blame the `count == -1` line in vim but got stuck at 7.0.001 and I was to lazy to look up the archival repository. I suppose this was once used somewhere but that got deleted. Let's say assuming count >= 0 is fine.

**Comment by bfredl** [comment]
LGTM (sorry if last comments was too ambiguos, do what seems most reasonable). There is some unrelated changes left, though as most of these _once_ _were_ lint changes to relevant changes, I would say it is ok to keep them.   But as this commit does quite many changes to a core part of nvim I would like someone else to do a complete review as well.

**Comment by justinmk** [comment]
> I tried to blame the count == -1 line in vim but got stuck at 7.0.001 and I was to lazy to look up the archival repository.

@bfredl Does such a thing exist? I mean the Vim SVN (CVS?) repo from before the switch to Mercurial.

**Comment by bfredl** [comment]
vim.org talks about CVS repo but I haven't checked if the server is still alive...

**Comment by bfredl** [comment]
seems alive and goes back to 5-5-001, 16 yars ago

**Comment by justinmk** [comment]
Very cool. We should import that into a git repo and mirror it somewhere.

**Comment by cacplate** [comment]
Good, I'll wait for another review and keep up-to-date in the meantime.

**Comment by cacplate** [comment]
@bfredl I had a problem with this PR and had to redo some of the work. When you find the time, I'd appreciate if you could look at it one more time.
Thanks for your time.

**Comment by bfredl** [comment]
Yes, will look into it soon.

File: src/nvim/ops.c [hunk 45]
```diff
@@ -2512,10 +2504,11 @@ static void op_yank_reg(oparg_T *oap, bool message, yankreg_T *reg, bool append)
   return;
 }
 
-static void yank_copy_line(yankreg_T *reg, struct block_def *bd, long y_idx)
+static void yank_copy_line(yankreg_T *reg, struct block_def *bd, size_t y_idx)
 {
-  char_u *pnew = xmallocz(bd->startspaces + bd->endspaces + bd->textlen);
-
+  assert(bd->startspaces + bd->endspaces + bd->textlen >= 0);
```
Comment (bfredl) [inline src/nvim/ops.c]: can any of these be negative? (the conversions below seem to suggest otherwise)

File: src/nvim/ops.c [hunk 46]
```diff
@@ -2926,8 +2921,9 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
       memset(ptr, ' ', (size_t)bd.endspaces);
       ptr += bd.endspaces;
       /* move the text after the cursor to the end of the line. */
+      assert(oldlen <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/ops.c]: unnecessary as `oldlen` is the length of an (actual) line.

File: src/nvim/ops.c [hunk 47]
```diff
@@ -5519,8 +5511,9 @@ static bool get_clipboard(int name, yankreg_T **target, bool quiet)
     reg->y_type = kMTUnknown;
   }
 
-  reg->y_array = xcalloc(lines->lv_len, sizeof(uint8_t *));
-  reg->y_size = lines->lv_len;
+  assert(lines->lv_len >= 0);
```
Comment (bfredl) [inline src/nvim/ops.c]: I think it safe to assume the length of a list is nonnegative. (we could have asserts in the functions for manipulating lists, but not in every case a list is just being used).

**Comment by bfredl** [comment]
some comments otherwise LGTM. maybe @justinmk or someone wants to to take a quick look, otherwise let's merge this.

File: src/nvim/ops.c (hunk 45 above)
Comment (cacplate) [inline src/nvim/ops.c]: indeed. A better way would be to assert the three separately but I think it's safe without. Removing this one.

**Comment by cacplate** [comment]
@justinmk sorry to bump. Can we merge this? Unless you think I need to revise something.
